### DISCUSSION TITLE
[Feature] Request timeout

### DISCUSSION THREAD

**Discussion by luisaveiro** [wolfpackthatcodes/typescript-http-client#42] [Ideas]
#### Summary

Provide a `timeout` method that specifies the maximum number of milliseconds to wait for a response. 

---

#### Possible implementation

We can provide the `AbortSignal:timeout()` static method in the fetch() API options. The method returns an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) that will automatically abort after a specified time.

##### Code example

```typescript
fetch(url, { signal: AbortSignal.timeout(5000) })
````

For more information about `AbortSignal:timeout()` static method, read [mdm web docs](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static).

##### Method signature

We can have a `timeout` method that accepts the maximum number of milliseconds the request should wait for a response.

```typescript
export default class HttpClient {
  /**
   * Specify the timeout (in milliseconds) for the pending request.
   *
   * @param {number} milliseconds
   *
   * @returns {this}
   */
  public timeout(milliseconds: number): this {
    // code ...
  }
}
```

---

#### Code examples

```typescript
import { HttpClient } from '@wolfpackthatcodes/http-client';

const response = new HttpClient()
  .timeout(3000)
  .get('https://api.example.local/users');
```

**Comment by ageddesi** [comment]
Would we want to supply a default in the class or set it so that the class is 0 to begin with and if that is the number present, when building up the config it ignores adding a AbortSignal. or should the function also toggle a flag as well as setting the milliseconds.

**Comment by luisaveiro** [comment]
@ageddesi there is additional functionality I want to achieve with this feature. I want to give the option for developers to manually abort the request. According to [mdm web docs](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal#aborting_a_fetch_with_timeout_or_explicit_abort), you can abort a fetch with timeout or explicit abort.

##### Code example

```typescript
const controller = new AbortController();
const timeoutSignal = AbortSignal.timeout(5000);

const res = await fetch(url, {
  // This will abort the fetch when either signal is aborted
  signal: AbortSignal.any([controller.signal, timeoutSignal]),
});
```

---

#### Functionality

The functionality we want to achieve for the HTTP Client are as follows:

- The HTTP Client should include the `AbortController()` by default.
- Provide a `timeout(milliseconds: number)` method to have the request abort within the timeout.
- Provide an `abort()` method to manually abort the request.
- The HTTP Client should provide two new Exceptions (`RequestTimeoutException`, `RequestAbortException`).

##### Code example

````typescript
import { HttpClient } from '@wolfpackthatcodes/http-client';

const httpClient = new HttpClient()
  .timeout(3000);
  
abortBtn.addEventListener("click", () => {
  httpClient.abort();
});

refreshBtn.addEventListener("click", () => {
  httpClient.get('https://api.example.local/users');
});
````

**Comment by luisaveiro** [comment]
While implementing this feature, there is a limitation of [browser compatibility](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/any_static#browser_compatibility) for the `AbortSignal.any()` static method. This method isn't fully supported across browsers, and only Node.js v20.3.0 has full support.

Due to this limitation, we cannot combine the `AbortSignal.timeout()` and `AbortController().signal` without utilising the `AbortSignal.any()` method.

**Comment by luisaveiro** [comment]
I have created a PR https://github.com/wolfpackthatcodes/typescript-http-client/pull/133
